### DISCUSSION TITLE
rook-ceph performance tuning

### DISCUSSION THREAD

**Discussion by Jiawei0227** [rook/rook#9253] [Q&A] [answered]
Hey guys, I am running rook-ceph on my cluster and I am trying to find the best practice for configuring parameters to use for my cluster. In general, I want to understand which parameters should I try to tune to get the maximum performance with limited resources.

My environment: a 3 node cluster with 4 NVMe data disks of size 3.8T. Each node has 28 core(I am supposing 48vCPU) and 768G Memory. I only want storage related pods to take about 20-25% compute resources(CPU and Memory)

The parameters I am tuning right now are:

- osdPerDevice
- mon, osd, mgr resources limit

In general, I find more osdPerDevice on NVMe seems to get a slightly better performance? Wondering if anyone has any insight on the performance tuning here that can help me understand:

1. What other parameters should I try to tune?
2. What's the best practice I should use? Is 3.84T NVMe normally used as 1 OSD or 2 OSD?
3. Is there a performance QoS limit on a Ceph object store or on an object bucket? I notice that even if I add more resources for OSD and MON it does not increase the performance for stress test on a single bucket.
4. It seems RGW also takes a lot of memory. Wondering if there is a parameter I can tune there? I try to avoid memory limits as it will kill the RGW if it goes OOM. 
5. I am mostly concerned the memory usage since I dont want to add limit to the pods. But I am afraid the components will use crazy amount of memory in some circumstances, just wondering if I can add some threshold somewhere.

Any help is appreciated!

**Comment by leseb** [comment, marked as answer]
All of these are good questions but they will find more responses if they are on the [Ceph mailing list ](https://ceph.io/en/community/connect/). Also, keep in mind that Ceph has reasonable default values to accommodate most workloads. Each time there is a value that Ceph developers believe can increase the performance, they change it so everyone benefits from it. It is a little bit of a rabbit hole once we start diving into fine-tuning Ceph (or any distributed system).
A good general rule is to apply resource limits to a pod, even if this could cause OOM, Rook-Ceph has the internals to configure Ceph memory when resource limits are applied, so nothing to tune here.
If you have an NVMe disk, it's generally a good idea to collocate a few OSDs, up to 5 depending on the model and the risk you are willing to take over performance (if you lose the disk you lose N OSDs).

Some limits you could apply:

* mon: memory 4GiB, CPU 1
* mgr: memory 8GiB, CPU 1 (depends on how many modules are enabled, the more module the more memory
* osd: memory: 16GiB, CPU 3, with 3 OSDs per NVMe

Obviously, take this with a grain of salt, these are just baselines, if you have more resources to allocate they do it, it's always better.
